cluster-autoscaler/cloudprovider/gce/autoscaling_gce_client.go

```diff
@@ -274,32 +270,31 @@ func instanceIdsToNamesMap(instanceProviderIds []string) map[string]bool {
 	return instanceNames
 }
 
-func (client *autoscalingGceClientV1) waitForOp(operation *gce.Operation, project, zone string, isDeletion bool) error {
-	pollInterval := client.operationPollInterval
-	if isDeletion {
-		pollInterval = client.operationDeletionPollInterval
-	}
-	for start := time.Now(); time.Since(start) < client.operationWaitTimeout; time.Sleep(pollInterval) {
-		klog.V(4).Infof("Waiting for operation %s %s %s", project, zone, operation.Name)
-		registerRequest("zone_operations", "get")
-		if op, err := client.gceService.ZoneOperations.Get(project, zone, operation.Name).Do(); err == nil {
-			klog.V(4).Infof("Operation %s %s %s status: %s", project, zone, operation.Name, op.Status)
-			if op.Status == "DONE" {
-				if op.Error != nil {
-					errBytes, err := op.Error.MarshalJSON()
-					if err != nil {
-						errBytes = []byte(fmt.Sprintf("operation failed, but error couldn't be recovered: %v", err))
-					}
-					return fmt.Errorf("error while getting operation %s on %s: %s", operation.Name, operation.TargetLink, errBytes)
-				}
+func (client *autoscalingGceClientV1) WaitForOperation(operationName, operationType, project, zone string) error {
+	ctx, cancel := context.WithTimeout(context.TODO(), client.operationWaitTimeout)
```

Comment:
I don't think the cancel works - it's only called in the defer when exiting the function, but we'd never reach the exit of the function if we e.g. got stuck in the for loop below. The cancel should be issued after `client.operationWaitTimeout` passes since entering the method, you can use `time.After(client.operationWaitTimeout)` for that.

The cleanest solution that comes to mind would be to extract the body of the for loop into a separate function that takes a channel, and returns the status and error through that channel. Start that function in a goroutine here, and then do a `select` on either the result channel (in which case either return if DONE/err!=nil, or trigger another call of the extracted function otherwise), or the channel returned from `time.After` (in which case call cancel and return). WDYT?

Reply:
The cancel function is here only to release the resources, all the timeouts are being handled inside `context.WithTimeout` as they are wrapping parent context into a new one with timer which executes cancel function by itself after some period of time.

I've also checked the internals of `Do()` call and they are also accounting for cancelled channel, so I think we are safe from this perspective. We of course may add select statement with `time.After` and `ctx.Done` and it shouldn't affect performance anyhow, but it's redundant from my perspective

Reply:
You're right, I misunderstood the `WithTimeout` method. This should work, could you just change `context.TODO()` to `context.Background()`? In this case we want a fresh, empty context, `TODO()` is more for cases where some context should actually be propagated to this function, but isn't yet.

Reply:
Totally makes sense, changing it from TODO to Background. I personally would expect all the client methods to accept context as the first parameter, but it's probably the topic for the other PR